gemrb/tests/INIImporter/Test_INIImporter.cpp

```diff
@@ -0,0 +1,132 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include <gtest/gtest.h>
+
+#include "../../core/Streams/FileStream.h"
+#include "../../plugins/INIImporter/INIImporter.h"
+
+namespace GemRB {
+
+static const path_t SAMPLE_FILE = PathJoin("tests", "resources", "INIImporter", "sample.ini");
+
+class INIImporter_Test : public testing::Test {
+protected:
+	FileStream* stream;
+	INIImporter unit;
+public:
+	void SetUp() override {
+		this->stream = new FileStream{};
+
+		stream->Open(SAMPLE_FILE);
+		assert(unit.Open(stream));
+		// INIImporter deletes stream internally after use
+		this->stream = nullptr;
+	}
+};
+
+TEST_F(INIImporter_Test, GetTagsCount) {
+	EXPECT_EQ(unit.GetTagsCount(), 4);
+}
+
+TEST_F(INIImporter_Test, GetTagNameByIndex) {
+	EXPECT_EQ(StringView{"sectionA"}, unit.GetTagNameByIndex(0));
+	EXPECT_EQ(StringView{"sectionB"}, unit.GetTagNameByIndex(1));
+	EXPECT_EQ(StringView{"sectionC"}, unit.GetTagNameByIndex(2));
+	EXPECT_EQ(StringView{"sectionC"}, unit.GetTagNameByIndex(3));
+
+	EXPECT_EQ(StringView{}, unit.GetTagNameByIndex(200));
+}
+
+TEST_F(INIImporter_Test, GetKeysCount) {
+	EXPECT_EQ(unit.GetKeysCount(StringView{"SectionA"}), 6);
+	EXPECT_EQ(unit.GetKeysCount(StringView{"sectiona"}), 6);
+	EXPECT_EQ(unit.GetKeysCount(StringView{"SectionB"}), 1);
+	// first tag match
+	EXPECT_EQ(unit.GetKeysCount(StringView{"SectionC"}), 1);
+
+	EXPECT_EQ(unit.GetKeysCount(StringView{"nope"}), 0);
+}
+
+TEST_F(INIImporter_Test, GetKeyNameByIndex) {
+	auto sectionA = StringView{"SectionA"};
+	auto sectionB = StringView{"SectionB"};
+	auto sectionC = StringView{"SectionC"};
+
+	EXPECT_EQ(StringView{"stringValue"}, (unit.GetKeyNameByIndex(sectionA, 0)));
+	EXPECT_EQ(StringView{"intValue"}, (unit.GetKeyNameByIndex(sectionA, 1)));
+	EXPECT_EQ(StringView{"intValue"}, (unit.GetKeyNameByIndex(sectionA, 2)));
+	EXPECT_EQ(StringView{"floatValue"}, (unit.GetKeyNameByIndex(sectionA, 3)));
+	EXPECT_EQ(StringView{"trueValue"}, (unit.GetKeyNameByIndex(sectionA, 4)));
+	EXPECT_EQ(StringView{"anotherStringValue"}, (unit.GetKeyNameByIndex(sectionB, 0)));
+	EXPECT_EQ(StringView{"iAm"}, (unit.GetKeyNameByIndex(sectionC, 0)));
+
+	EXPECT_EQ(StringView{}, (unit.GetKeyNameByIndex(sectionA, 100)));
+	EXPECT_EQ(StringView{}, (unit.GetKeyNameByIndex(StringView{"section404"}, 0)));
+}
+
+TEST_F(INIImporter_Test, GetKeyAsString) {
+	auto sectionA = StringView{"SectionA"};
+
+	EXPECT_EQ(StringView{"abc"}, (unit.GetKeyAsString(sectionA, StringView{"stringValue"})));
+	EXPECT_EQ(StringView{"abc"}, (unit.GetKeyAsString(sectionA, StringView{"Stringvalue"})));
+	EXPECT_EQ(StringView{"n/a"}, (unit.GetKeyAsString(sectionA, StringView{"strongValue"}, StringView{"n/a"})));
+	// first match
+	EXPECT_EQ(StringView{"123"}, (unit.GetKeyAsString(sectionA, StringView{"intValue"})));
```

Comment:
I don't find this terribly intuitive — I'd expect the second key to override the value.

Reply:
Agree, needs some rework though

Reply:
Ok, done. But only for keys inside of the same section. Later sections of the same name don't merge anything so they are available via iteration but not by direct query.

Reply:
That sounds treacherous. I think it should behave the same as keys, overriding any previous section (not merging).